### ISSUE TITLE
OpTypeRuntimeArray should only be used certain purposes

### ISSUE THREAD

**Issue by dneto0** [KhronosGroup/SPIRV-Tools#1458]
Instruction description of OpTypeRuntimeArray says "Objects of this type can only be created with OpVariable using the StorageBuffer or Uniform Storage Classes."

SPV_EXT_descriptor_indexing says this is superceded by Vulkan environment spec that says

"OpTypeRuntimeArray must only be used for:
the last member of an OpTypeStruct that is in the StorageBuffer storage class decorated as Block, or that is in the Uniform storage class decorated as BufferBlock."

This is so runtime array can be used for an unbounded size list of descriptors. See RuntimeDescriptorArrayEXT in SPV_EXT_descriptor_indexing

**Comment by dneto0** [comment]
Also, it's reasonable to say that "instantiating" an aggregate value also implicitly "instantiates" all the aggregate's compojnents/members.  So it's invalid for a load to return a struct whose last member is a runtime array.

cc: @johnkslang to see if he agrees with my assumption that "instantiating an aggregate implicitly instantiates its members or components"

**Comment by johnkslang** [comment]
Agreed this is certainly the intent (we would not intend to say you cannot load X, but if X is a member of a struct, then actually you can load it, via loading the struct).  So, worst case here is a change to the SPIR-V spec to make that more clear. For validiation, it should be invalid to load a runtime-sized array either directly or indirectly.
